[RNMobile] Refactor `react-native-editor` initialization test

fluiddot opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Add test case for covering that the editor initializes correctly and renders [the example initial HTML content](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/src/initial-html.js).

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
1. Observe that PR check `Unit Tests / Mobile` passes.
2. Run command `npm run native test` and observe that all tests pass.

## Screenshots <!-- if applicable -->
N/A

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Improvement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [x] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

fluiddot committed:
Refactor react-native-editor initialization test

fluiddot commented on packages/react-native-editor/src/test/index.test.js:
```diff
@@ -27,158 +29,202 @@ const initGutenberg = ( registerParams ) => {
-		const { getByTestId } = initGutenberg();
-		const blockList = waitFor( () => getByTestId( 'block-list-wrapper' ) );
-		expect( blockList ).toBeDefined();
+			expect( blockList ).toHaveProperty( 'type', 'View' );
```
As a workaround for preventing adding false-positives as it was outlined [here](https://github.com/WordPress/gutenberg/pull/37921/files#r783288889), instead of checking the existence, we check that the component type matches what we expect.

dcalhoun replied:
Interesting technique. 

It is interesting that [RNTL's asynchronous examples](https://callstack.github.io/react-native-testing-library/docs/react-navigation#example-tests) do showcase utilizing `toBeTruthy` when it can easily produce false positives if one forgets to use `await`. 

From reviewing Testing Library's [`findBy` guides](https://testing-library.com/docs/example-findByText), they recommend [leveraging `toBeVisible`](https://github.com/testing-library/jest-dom/tree/v5.16.1#tobevisible), which is a `jest-dom` matcher. Unfortunately, the sibling `jest-native` package does not have a [comparable matcher](https://github.com/testing-library/jest-native/tree/v4.0.4#matchers). 

I wonder if it is worthwhile to [extend `expect`](https://jestjs.io/docs/26.x/expect#expectextendmatchers) our own custom `toBeVisible` matcher to help avoid this common pitfall? WDYT?

fluiddot replied:
> It is interesting that RNTL's asynchronous examples do showcase utilizing toBeTruthy when it can easily produce false positives if one forgets to use await.
Yep, I think in general `toBeTruthy` should work for this type of expectation, but I'd lean towards a more robust solution like using a custom expect as `toBeVisible` 👍 .

> From reviewing Testing Library's [`findBy` guides](https://testing-library.com/docs/example-findByText), they recommend [leveraging `toBeVisible`](https://github.com/testing-library/jest-dom/tree/v5.16.1#tobevisible), which is a `jest-dom` matcher. Unfortunately, the sibling `jest-native` package does not have a [comparable matcher](https://github.com/testing-library/jest-native/tree/v4.0.4#matchers).
> 
> I wonder if it is worthwhile to [extend `expect`](https://jestjs.io/docs/26.x/expect#expectextendmatchers) our own custom `toBeVisible` matcher to help avoid this common pitfall? WDYT?

That's very interesting, thanks for sharing it ❤️ . I really like the idea of implementing a matcher like that one, I'll review the implementation of [`toBeVisible`](https://github.com/testing-library/jest-dom/tree/v5.16.1#tobevisible), in case we can provide implementation as close as possible to its original.

dcalhoun replied:
To be clear, I am happy to merge this assertion as-is. We could always follow up with a `toBeVisible` query at a later time if you'd prefer.

fluiddot replied:
> To be clear, I am happy to merge this assertion as-is. We could always follow up with a `toBeVisible` query at a later time if you'd prefer.

Ok, to be honest, it makes sense to me to tackle this in a different PR. Besides, on that PR we could even try to update all tests with that expectation 👍 .

fluiddot replied:
@dcalhoun I explored [`@testing-library/jest-native`](https://github.com/testing-library/jest-native) package and [`toBeVisible` matcher](https://github.com/testing-library/jest-dom/blob/main/src/to-be-visible.js) and decided to implement a RN version in [this PR](https://github.com/testing-library/jest-native/pull/73). It's still in draft but I'd appreciate it if you could take a look and let me know your opinion, thanks 🙇 !

fluiddot commented on packages/react-native-editor/src/test/index.test.js:
```diff
@@ -27,158 +29,202 @@ const initGutenberg = ( registerParams ) => {
-		expect( blockList ).toBeDefined();
+			expect( blockList ).toHaveProperty( 'type', 'View' );
+			expect( console ).toHaveLogged( 'Hermes is: true' );
+			// It's expected that some blocks are upgraded and inform about it (example: "Updated Block: core/cover")
+			expect( console ).toHaveInformed();
```
This might be giving us a clue about using outdated versions of blocks in [`initialHtml` file](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/src/initial-html.js).

dcalhoun replied:
By not passing arguments to `toHaveInformed`, I would imagine we expose ourselves to additional `console.info` calls accumulating without our knowledge. From reviewing the source of `@wordpress/jest-console`, I suppose there is no way to further constrain this assertion without passing the entirety of the explicit block update output, which is not manageable unfortunately. 



> This might be giving us a clue about using outdated versions of blocks in [`initialHtml` file](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/src/initial-html.js).

With this are you implying that we should fix the `initialHtml` rather than suppressing the `console.info` message here?

fluiddot replied:
> By not passing arguments to toHaveInformed, I would imagine we expose ourselves to additional console.info calls accumulating without our knowledge. From reviewing the source of @wordpress/jest-console, I suppose there is no way to further constrain this assertion without passing the entirety of the explicit block update output, which is not manageable unfortunately.

Exactly, there might be some cases where you could overlook extra or unexpected info logs using this function without passing arguments. In the case of this test, although we know the expected log output, I decided not to include it due to its size.

> With this are you implying that we should fix the initialHtml rather than suppressing the console.info message here?

Yep, I haven't investigated this further, but the fact that some blocks are being upgraded makes me wonder whether they should be reviewed, in case the HTML we're using for them is old.

fluiddot replied:
In any case, I think we could address this in a different PR, I'll open a ticket for this as a follow-up.

fluiddot commented on packages/react-native-editor/src/test/index.test.js:
```diff
@@ -7,18 +7,20 @@ import { render, waitFor } from 'test/helpers';
  */
+import { getBlockTypes, unregisterBlockType } from '@wordpress/blocks';
 import * as wpHooks from '@wordpress/hooks';
+import '@wordpress/jest-console';
```
I found this library quite useful for assuring we don't produce unexpected logs. In the future, it would be great to enable it globally 🤩 .

An important note about using it is that it silences potential logs that happen after the test. For example, [these test cases](https://github.com/WordPress/gutenberg/blob/e8d809fc565faad2ee3530d7cb261a6af511317c/packages/react-native-editor/src/test/index.test.js#L203-L228) are generating "act" warnings (i.e. `An update to EditorProvider inside a test was not wrapped in act(...).`) due to the fact that some store state updates are happening after the test is finished. In the beginning, I thought of this as an issue, but I think it's safe to omit them as they are happening after the test ends.

dcalhoun replied:
> I found this library quite useful for assuring we don't produce unexpected logs. In the future, it would be great to enable it globally 🤩 .

Agreed. I would love to enable this globally, we'll just need to address all the current warnings and errors in our test logs first. 

> An important note about using it is that it silences potential logs that happen after the test. 

Yeah, this has confused me before while debugging failing web tests, which do enable `@wordpress/jest-console` globally. In order to intentionally log information, one has to explicitly assert the log should occur. 

> For example, [these test cases](https://github.com/WordPress/gutenberg/blob/e8d809fc565faad2ee3530d7cb261a6af511317c/packages/react-native-editor/src/test/index.test.js#L203-L228) are generating "act" warnings (i.e. `An update to EditorProvider inside a test was not wrapped in act(...).`) due to the fact that some store state updates are happening after the test is finished. In the beginning, I thought of this as an issue, but I think it's safe to omit them as they are happening after the test ends.

While these particular updates may not dangerous for this specific context, I think it may be dangerous to set the precedent that it is safe to ignore the warnings. It is very possible for a post-assertion update to nullify the validity of a successful assertion. E.g. if one asserts an element is present, a later state update may unexpectedly and erroneously removes the element, which would mean the test is a false positive. The logs occurring "after the test ends" I believe is merely a result of them occurring after the final assertion, which doesn't inherently make them safe. 

Is there a way for us to address the `act` warnings in these tests?

fluiddot replied:
> Agreed. I would love to enable this globally, we'll just need to address all the current warnings and errors in our test logs first.

In fact, I'm planning to open a ticket specific to addressing this topic.

> While these particular updates may not dangerous for this specific context, I think it may be dangerous to set the precedent that it is safe to ignore the warnings. It is very possible for a post-assertion update to nullify the validity of a successful assertion. E.g. if one asserts an element is present, a later state update may unexpectedly and erroneously remove the element, which would mean the test is a false positive. The logs occurring "after the test ends" I believe is merely a result of them occurring after the final assertion, which doesn't inherently make them safe.

Good point, I understand and share your concern about the potential side effects of ignoring the warnings. However, using the `@wordpress/jest-console` package would make it very easy to ignore them, so it would be great to discuss whether including (although its benefits) it would be safe in relation to this issue. I guess the main problem here is how to know when the editor and its components finish triggering all the state updates, although I'd like to note that it's also important that we identify potential errors upon unmounting the editor or components, in order to prevent memory leaks caused by unclosed listeners.

Being this said, I'm thinking of creating a native version of [the supported matchers](https://github.com/WordPress/gutenberg/blob/trunk/packages/jest-console/src/supported-matchers.js) and excluding the error logs. This way we would assure the "act" errors (note that they're warnings but logged via `console.error`) are logged hence, we could easily identify them in the output.

> Is there a way for us to address the act warnings in these tests?

Actually, I managed to find a workaround by delaying the query passed to the `waitFor` function an execution block cycle (check the [`rnmobile/add/use-immediate-wait-for` branch](https://github.com/WordPress/gutenberg/compare/rnmobile/add/use-immediate-wait-for)). I haven't investigated the original issue, but I have the gut feeling that it might be related to how the store actions and updates are executed.

dcalhoun replied:
> However, using the `@wordpress/jest-console` package would make it very easy to ignore them, so it would be great to discuss whether including (although its benefits) it would be safe in relation to this issue. 

The absence of `@wordpress/jest-console` makes it far easier for warnings and errors to go unnoticed, as they will not necessarily cause test failures. I agree `@wordpress/jest-console` makes it easier to explicitly hide warnings and errors. I would posit the former is worst. The latter is at least an intentional act. 

> I guess the main problem here is how to know when the editor and its components finish triggering all the state updates, although I'd like to note that it's also important that we identify potential errors upon unmounting the editor or components, in order to prevent memory leaks caused by unclosed listeners.

Agreed. We appear to be aligned that it is not safe or good practice to ignore the warnings or assume they are safe. 

> Being this said, I'm thinking of creating a native version of [the supported matchers](https://github.com/WordPress/gutenberg/blob/trunk/packages/jest-console/src/supported-matchers.js) and excluding the error logs. This way we would assure the "act" errors (note that they're warnings but logged via `console.error`) are logged hence, we could easily identify them in the output.

Maybe I am mistaken, but I interpret the intent of `@wordpress/jest-console` is to _fail tests_ when console logs, warnings, or errors occur. `act` errors would trigger test failures when `@wordpress/jest-console` is enabled. So, `@wordpress/jest-console` itself should already accomplish your proposed end goal: ensure `act` errors are not overlooked. I.e. someone is far more likely to notice — and hopefully fix — a failing test (due to an `act` error) than a passing test that happens to also log an `act` error. 

<details><summary>Example Failure from act + @wordpress/jest-console</summary>

```
 FAIL  ../block-library/src/cover/test/edit.native.js (7.967 s)
  when no media is attached
    ○ skipped adds an image or video
  when an image is attached
    ✕ discards canceled focal point changes (343 ms)
    ○ skipped edits the image
    ○ skipped replaces the image
    ○ skipped clears the image within image edit button
    ○ skipped toggles a fixed background
    ○ skipped edits the focal point with a slider
    ○ skipped edits the focal point with a text input
    ○ skipped clears the media within cell button
  color settings
    ○ skipped sets a color for the overlay background when the placeholder is visible
    ○ skipped sets a gradient overlay background when a solid background was already selected
    ○ skipped toggles between solid colors and gradients
    ○ skipped clears the selected overlay color and mantains the inner blocks

  ● when an image is attached › discards canceled focal point changes

    expect(jest.fn()).not.toHaveErrored(expected)

    Expected mock function not to be called but it was called with:
    ["Warning: An update to %s inside a test was not wrapped in act(...).·
    When testing, code that causes React state updates should be wrapped into act(...):·
    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */·
    This ensures that you're testing the behavior the user would see in the browser. Learn more at https://reactjs.org/link/wrap-tests-with-act%s", "ForwardRef(NavigationContainer)", "
        at NavigationContainer (/Users/[REDACTED]/Sites/a8c/gutenberg/node_modules/@react-navigation/native/lib/commonjs/NavigationContainer.tsx:40:5)
        at View
        at View (/Users/[REDACTED]/Sites/a8c/gutenberg/node_modules/react-native/jest/mockComponent.js:28:18)
        at BottomSheetNavigationContainer (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/components/src/mobile/bottom-sheet/bottom-sheet-navigation/navigation-container.native.js:63:2)
        at View
        at View (/Users/[REDACTED]/Sites/a8c/gutenberg/node_modules/react-native/jest/mockComponent.js:28:18)
        at View
        at View (/Users/[REDACTED]/Sites/a8c/gutenberg/node_modules/react-native/jest/mockComponent.js:28:18)
        at KeyboardAvoidingView (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/components/src/mobile/bottom-sheet/keyboard-avoiding-view.native.js:26:16)
        at Modal
        at /Users/[REDACTED]/Sites/a8c/gutenberg/test/native/setup.js:88:8
        at BottomSheet (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/components/src/mobile/bottom-sheet/index.native.js:51:16)
        at BottomSheetSettings (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/block-editor/src/components/block-settings/container.native.js:29:2)
        at WithDispatch(BottomSheetSettings) (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/data/src/components/with-dispatch/index.js:95:26)
        at /Users/[REDACTED]/Sites/a8c/gutenberg/packages/data/src/components/with-select/index.js:56:24
        at WithSelect(WithDispatch(BottomSheetSettings)) (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/compose/src/higher-order/pure/index.tsx:35:3)
        at SlotFillProvider (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/components/src/slot-fill/provider.js:18:16)
        at CoverEdit"],["Warning: An update to %s inside a test was not wrapped in act(...).·
    When testing, code that causes React state updates should be wrapped into act(...):·
    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */·
    This ensures that you're testing the behavior the user would see in the browser. Learn more at https://reactjs.org/link/wrap-tests-with-act%s", "Cover", "
        at Cover (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/block-library/src/cover/edit.native.js:83:2)
        at WithDispatch(Component) (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/data/src/components/with-dispatch/index.js:95:26)
        at /Users/[REDACTED]/Sites/a8c/gutenberg/packages/data/src/components/with-select/index.js:56:24
        at WithSelect(WithDispatch(Component)) (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/compose/src/higher-order/pure/index.tsx:35:3)
        at Edit (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/block-editor/src/components/block-edit/edit.native.js:29:10)
        at WithToolbarControls(Edit) (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/block-editor/src/hooks/align.js:119:11)
        at WithInspectorControl(WithToolbarControls(Edit)) (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/block-editor/src/hooks/anchor.js:68:45)
        at WithToolbarControls(WithInspectorControl(WithToolbarControls(Edit))) (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/block-editor/src/hooks/style.js:262:33)
        at WithFilters(Edit) (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/components/src/higher-order/with-filters/index.js:49:18)
        at BlockEdit (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/block-editor/src/components/block-edit/index.js:23:10)
        at SlotFillProvider (/Users/[REDACTED]/Sites/a8c/gutenberg/packages/components/src/slot-fill/provider.js:18:16)
        at CoverEdit"]

      34 |      function assertExpectedCalls() {
      35 |              if ( spy.assertionsNumber === 0 && spy.mock.calls.length > 0 ) {
    > 36 |                      expect( console ).not[ matcherName ]();
         |                      ^
      37 |              }
      38 |      }
      39 |

      at Object.assertExpectedCalls (packages/jest-console/src/index.js:36:4)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 12 skipped, 13 total
Snapshots:   0 total
Time:        9.12 s
Ran all test suites matching /packages\/block-library\/src\/cover\/test\/edit.native.js/i.
```

</details>

> Actually, I managed to find a workaround by delaying the query passed to the `waitFor` function an execution block cycle (check the [`rnmobile/add/use-immediate-wait-for` branch](https://github.com/WordPress/gutenberg/compare/rnmobile/add/use-immediate-wait-for)). I haven't investigated the original issue, but I have the gut feeling that it might be related to how the store actions and updates are executed.

Interesting. I have limited knowledge of the error, so I am uncertain to the appropriateness of the possible fix. 🤔

fluiddot replied:
> Maybe I am mistaken, but I interpret the intent of `@wordpress/jest-console` is to _fail tests_ when console logs, warnings, or errors occur. `act` errors would trigger test failures when `@wordpress/jest-console` is enabled. So, `@wordpress/jest-console` itself should already accomplish your proposed end goal: ensure `act` errors are not overlooked. I.e. someone is far more likely to notice — and hopefully fix — a failing test (due to an `act` error) than a passing test that happens to also log an `act` error.

That's correct, however, it doesn't identify the `act` warnings that happen after the test finishes, which are the ones that we wouldn't notice if we enable `@wordpress/jest-console`.

Let me share the following table, in order to make clearer the benefits and drawbacks of whether enable `@wordpress/jest-console`:
Using `@wordpress/jest-console`|Not using `@wordpress/jest-console`
--|--
| **Benefits:**<br>- `act` warnings will make the test fail hence, it will require contributors to fix them.<br>- Other log types will also be required to be addressed, which would help in having more robust tests.<br><br>**Drawbacks:**<br>- Any type of log that happens before or after a test won't be displayed and either considered a test failure. This implies that for component rendering, some `act` warnings will be unnoticed.  | **Benefits:**<br>- We won't experience the drawbacks of using `@wordpress/jest-console` related to omitting warnings or error logs out of the scope of tests.<br><br>**Drawbacks:**<br>- Nevertheless, the logs won't produce test failures, which would require manual observation of test output.<br>- For some tests, the output might be too verbose, unless we mock `console` functions for those cases. |

fluiddot replied:
> Interesting. I have limited knowledge of the error, so I am uncertain to the appropriateness of the possible fix. 🤔

I left those changes on hold until I manage to have some spare time to do a deeper investigation, but as far as I tested, looks like the workaround addresses the issue.

fluiddot replied:
@dcalhoun I removed the use of `@wordpress/jest-console` package in https://github.com/WordPress/gutenberg/pull/37955/commits/4e083313fb425ba0ee5968ab5af05fdaffd4ea60, note that it required a small refactor to cover the same logic. I'd appreciate it if you could take a look and let me know your thoughts about that approach, thanks 🙇 .

fluiddot replied:
Looks like using Jest v27 would help on this issue, per [this comment](https://github.com/WordPress/gutenberg/pull/33287#issuecomment-1015399942), I noticed that a recently added test is failing due to an `act` warning being logged after the test is finished 🤩 , and includes the `@wordpress/jest-console` (this is the first test we have introduced it).

dcalhoun replied:
> That's correct, however, it doesn't identify the `act` warnings that happen after the test finishes, which are the ones that we wouldn't notice if we enable `@wordpress/jest-console`.

Interesting. I was not aware of that. That is really surprising. I wonder why that is. My understanding of `act` warnings was that they specifically warn about changes that occur after the test finishes. `@wordpress/jest-console` "swallows" all act warnings — both during and after tests — but only fails the test for act warnings that occur "during" the test? 

> @dcalhoun I removed the use of `@wordpress/jest-console` package in [4e08331](https://github.com/WordPress/gutenberg/commit/4e083313fb425ba0ee5968ab5af05fdaffd4ea60), note that it required a small refactor to cover the same logic. I'd appreciate it if you could take a look and let me know your thoughts about that approach, thanks 🙇 .

I think the new approach is an appropriate solution while we further investigate more global, long-term solutions. 

> Looks like using Jest v27 would help on this issue, per [this comment](https://github.com/WordPress/gutenberg/pull/33287#issuecomment-1015399942), I noticed that a recently added test is failing due to an `act` warning being logged after the test is finished 🤩 , and includes the `@wordpress/jest-console` (this is the first test we have introduced it).

👀 Following along that conversation and exploration as well. 👍🏻

fluiddot replied:
> Interesting. I was not aware of that. That is really surprising. I wonder why that is. My understanding of `act` warnings was that they specifically warn about changes that occur after the test finishes. `@wordpress/jest-console` "swallows" all act warnings — both during and after tests — but only fails the test for act warnings that occur "during" the test?

Exactly, only fails for `act` warnings that occur during the test. However, when using Jest v27, I noticed that this behavior changed and also makes the test fail when those warnings happen after the test 🤷‍♂️ .

> I think the new approach is an appropriate solution while we further investigate more global, long-term solutions.

Great, I'll proceed with it, and we'll see how we could improve it in the next iteration 👍.

fluiddot replied:
Heads up that I re-enabled this import as with Jest v27 the `act` warnings that occur after the test are also identified as errors.

fluiddot commented on packages/react-native-editor/src/test/index.test.js:
```diff
@@ -27,158 +29,202 @@ const initGutenberg = ( registerParams ) => {
-	return render( <EditorComponent /> );
+	let screen;
+	// This guarantees that setup module is imported on every test, as it's imported upon Editor component rendering.
+	jest.isolateModules( () => {
+		screen = render( <EditorComponent { ...editorProps } /> );
+	} );
```
With this approach, we no longer need to reset modules for testing the call order of hooks and callbacks. Besides, I found out that resetting modules was breaking the test cases that render the editor.

fluiddot committed:
Remove commented line code

dcalhoun commented on packages/react-native-editor/src/test/index.test.js:
```diff
@@ -27,158 +29,202 @@ const initGutenberg = ( registerParams ) => {
-			};
+		it( 'sets up locale before editor is initialized', () => {
+			const mockOnModuleImported = jest.fn();
+			// jest.resetModules();
```
This appears to be an oversight and should be removed, correct? 
```suggestion
```

fluiddot replied:
Good catch, looks like I fixed it before setting the PR to review but unfortunately, I forgot to push [the commit](https://github.com/WordPress/gutenberg/pull/37955/commits/ae0aa9fcee296669fb3720c5681cf58d4ac78c41) 😓 , sorry for the noise.

dcalhoun commented on packages/react-native-editor/src/test/index.test.js:
```diff
@@ -27,158 +29,202 @@ const initGutenberg = ( registerParams ) => {
 
-		expect( beforeInitCallOrder ).toBeLessThan( onSetupImportedCallOrder );
-	} );
+			const blockList = await waitFor( () =>
+				getByTestId( 'block-list-wrapper' )
+			);
```
The usage of `waitFor` does not appear to be necessary. A synchronous query works. WDYT about avoiding asynchronous queries unless they are required? We could also remove the `async` decorator on the anonymous test function. 

```js
const blockList = getByTestId( 'block-list-wrapper' );
```

fluiddot replied:
Right, I agree with only providing asynchronous queries when required. I'll update all cases with your suggestion, thanks 🙇 !

fluiddot replied:
I applied this suggestion in https://github.com/WordPress/gutenberg/pull/37955/commits/b1964b5eedef8c3f5667947be81211189cc34a76.

dcalhoun commented on packages/react-native-editor/src/test/index.test.js:
```diff
@@ -27,158 +29,202 @@ const initGutenberg = ( registerParams ) => {
-		expect( hookName ).toBe( 'native.render' );
-		expect( hookCallOrder ).toBeGreaterThan( onRenderEditorCallOrder );
-	} );
+			const blockList = await waitFor( () =>
+				getByTestId( 'block-list-wrapper' )
+			);
```
The usage of `waitFor` does not appear to be necessary. A synchronous query works. WDYT about avoiding asynchronous queries unless they are required?  We could also remove the `async` decorator on the anonymous test function. 

```js
const blockList = getByTestId( 'block-list-wrapper' );
```

fluiddot replied:
I applied this suggestion in https://github.com/WordPress/gutenberg/pull/37955/commits/b1964b5eedef8c3f5667947be81211189cc34a76.

dcalhoun reviewed (commented):
Thank you for putting this work together. I left a few comments for consideration.

fluiddot committed:
Get block list component synchronously

fluiddot committed:
Remove jest-console package import

Due to this package removal the tests related to initialization have been refactored to cover the same logic.

fluiddot commented on packages/react-native-editor/src/test/index.test.js:
```diff
@@ -27,158 +28,218 @@ const initGutenberg = ( registerParams ) => {
+				.mockImplementation( jest.fn() );
+
+			const { getByTestId } = initGutenberg( {}, { initialData: '' } );
+			// Some of the store updates that happen upon editor initialization are executed at the end of the current
+			// Javascript block execution and after the test is finished. In order to prevent "act" warnings due to
+			// this behavior, we wait for the execution block to be finished before acting on the test.
+			await act(
+				() => new Promise( ( resolve ) => setImmediate( resolve ) )
+			);
```
If we don't wait for the JS block to be finished, we get a bunch of `act` warnings:
```
Warning: An update to EditorProvider inside a test was not wrapped in act(...).
```

dcalhoun replied:
Leaving a note that we are actively investigating this issue and approach further in #38052. We should likely retroactively apply whatever global solution we find there to this test.

dcalhoun reviewed (approved):
LGTM. I support merging this in its current state if you'd like. We can always revisit these particular tests whenever we finish the Jest 27 work in #38052. Thank you for improving these tests! 🙇🏻

fluiddot commented:
> LGTM. I support merging this in its current state if you'd like. We can always revisit these particular tests whenever we finish the Jest 27 work in #38052. Thank you for improving these tests! 🙇🏻

Thanks for the review 🙇 , now that we have a better workaround ([the one you introduced for fixing the `act` warnings of the Jest v27 update](https://github.com/WordPress/gutenberg/pull/38052#issuecomment-1016603101)), I'll update the PR and apply it here too.

fluiddot committed:
Merge branch 'trunk' into rnmobile/update/editor-initialization-tests

# Conflicts:
#	packages/react-native-editor/src/test/index.test.js

fluiddot committed:
Improve error handling in waitFor test helper

fluiddot committed:
Update register gutenberg tests

`initGutenberg` function now returns the Editor component instead of rendering the editor.

fluiddot committed:
Update initialize editor test case

Now the initialize editor case renders the actual Editor component including the example initial html.

fluiddot commented on packages/react-native-editor/src/test/index.test.js:
```diff
@@ -18,24 +19,18 @@ import setupLocale from '../setup-locale';
 };
 
 describe( 'Register Gutenberg', () => {
-	beforeEach( () => {
-		// We need to reset modules to guarantee that setup module is imported on every test.
-		jest.resetModules();
-	} );
```
This is no longer necessary as the setup module import is guaranteed by using the `jest.isolateModules` function when rendering the editor.

fluiddot committed:
Remove async from initialize editor test case

fluiddot commented on test/native/helpers.js:
```diff
@@ -122,7 +123,7 @@ export function waitFor(
 			if ( ! result ) {
 				reject(
-					`waitFor timed out after ${ timeout }ms for callback:\n${ cb }`
+					`waitFor timed out after ${ timeout }ms for callback:\n${ cb }\n${ lastError.toString() }`
```
While testing I realized that errors not directly related to timeout, like syntax errors, were omitted. For this reason, I updated this function by displaying now the last error that caused the timeout:

**Example 1 - Querying for an unexisting text:**
<img width="713" alt="Screenshot 2022-01-24 at 11 17 30" src="https://user-images.githubusercontent.com/14905380/150770619-b6dfb388-ad85-46a0-9c61-0e0e36268dfd.png">

**Example 2 - Reference error:**
<img width="729" alt="Screenshot 2022-01-24 at 12 00 45" src="https://user-images.githubusercontent.com/14905380/150771127-da6fdc31-f7d3-495b-a9fb-6cbf9b17c61f.png">

fluiddot reviewed (commented):
@dcalhoun I've updated the PR with the latest changes from `trunk`, including the Jest v27 updated, which implied applying some updates to the test cases. Additionally, I updated the `initializes the editor` test case to cover the actual rendering of the editor with [the example initial HTML](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/src/initial-html.js). Let me know if you could do another review, thanks 🙇 !

dcalhoun reviewed (approved):
Great work! 🚀
